pkg/services/pluginsintegration/plugincontext/base_plugincontext.go

```diff
@@ -0,0 +1,62 @@
+package plugincontext
+
+import (
+	"context"
+	"runtime"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/useragent"
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/services/auth/identity"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/adapters"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/pluginconfig"
+	"github.com/grafana/grafana/pkg/services/pluginsintegration/pluginstore"
+	"github.com/grafana/grafana/pkg/setting"
+)
+
+type BasePluginContextProvider interface {
+	// GetBasePluginContext returns a plugin context for the given plugin and user.
+	// It does not add DatasourceInstaceSettings or AppInstanceSettings
+	GetBasePluginContext(ctx context.Context, plugin pluginstore.Plugin, user identity.Requester) backend.PluginContext
+}
+
+func BaseProvideService(cfg *setting.Cfg, pluginRequestConfigProvider pluginconfig.PluginRequestConfigProvider) *BaseProvider {
+	return &BaseProvider{
+		cfg:                         cfg,
+		pluginRequestConfigProvider: pluginRequestConfigProvider,
+		logger:                      log.New("base.plugin.context"),
+	}
+}
+
+func newBaseProvider(cfg *setting.Cfg, pluginRequestConfigProvider pluginconfig.PluginRequestConfigProvider) *BaseProvider {
+	return BaseProvideService(cfg, pluginRequestConfigProvider)
+}
+
+type BaseProvider struct {
```

Comment:
I'm curious if we couldn't just put this behavior in the existing plugin context struct and just use the interface (defined by the datasource/service package) to depend on the method it needs?

Reply:
I believe this is the same than Marcus suggests above? The problem is that the PluginContext implementation depends on the Datasource service (and therefore its implementation). So if the Datasource Service depends on a interface, even if the interface is minimal, and the only implementation of that interface is in the PluginContext implementation we enter in a cycle dependency.